mepo: 0.4.2 -> 0.5

amjoseph-nixpkgs opened the pull request:
###### Description of changes

First commit is a simple version bump.

Second commit adds `gpsd` as an optional dependency and points mepo's scripts' `gpspipe` to the correct path.  It also makes silent reporting of the user's location to Mozilla and geoclue opt-in rather than opt-out; this issue has been [reported upstream](https://todo.sr.ht/~mil/mepo-tickets/41) to request a build flag so we don't have to use patches.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
mepo: 0.4.2 -> 0.5

amjoseph-nixpkgs commented:
Added commits which were formerly part of #189345.

McSinyx commented:
Path to GeoClue binaries need substitution
and IMHO Mozilla location service should be
conditionally enabled in the same manner as the others.

BTW you should rebase the commits for each to be meaningful
(GitHub suggestion should be read-only).

amjoseph-nixpkgs commented:
> Path to GeoClue binaries need substitution

Done: 3b8d39d136de50cfd3b0e3159cb0b2b318b67565

> and IMHO Mozilla location service should be conditionally enabled in the same manner as the others.

I don't understand what you mean... isn't `reportLocationToMozilla` conditionally enabling the mozilla location service?

> BTW you should rebase the commits for each to be meaningful (GitHub suggestion should be read-only).

I don't understand this either... what do you mean by meaningful?  And what GitHub suggestion?

McSinyx commented:
> isn't reportLocationToMozilla conditionally enabling the mozilla location service?

It is, apologies, I misread :sweat_smile:

> what do you mean by meaningful [commit]?

37ceaf7d4c5c1157b3eb51262d520fc0ec0cdb77 and 3b8d39d136de50cfd3b0e3159cb0b2b318b67565 are examples of commits that only make sense with the context of the discussion.  Each commit should make sense by itself should it be merged.

> what GitHub suggestion?

The one created 37ceaf7d4c5c1157b3eb51262d520fc0ec0cdb77.

amjoseph-nixpkgs commented:
> [37ceaf7](https://github.com/NixOS/nixpkgs/commit/37ceaf7d4c5c1157b3eb51262d520fc0ec0cdb77) and [3b8d39d](https://github.com/NixOS/nixpkgs/commit/3b8d39d136de50cfd3b0e3159cb0b2b318b67565) are examples of commits that only make sense with the context of the discussion. Each commit should make sense by itself should it be merged.

Oh, I see, you want me to squash the history to clean it up.  Sure, no problem.

> Each commit should make sense by itself should it be merged.

Right.  Just to be clear though, I don't think it is a good idea to merge the version bump without the user location protection.  I will squash the history into two commits, but I hope that either both of them or else neither of them is merged.

> The one created [37ceaf7](https://github.com/NixOS/nixpkgs/commit/37ceaf7d4c5c1157b3eb51262d520fc0ec0cdb77).

I believe that I cherry-picked that into this PR as 37ceaf7d4c5c1157b3eb51262d520fc0ec0cdb77.

Commit:
mepo: add gpsdSupport

Commit:
mepo: add reportLocationToMozilla?false flag

mepo 0.5 silently reports the user's location to Mozilla without
asking first.  Let's disable this by default.

This has been reported to upstream with a request for a build-time
flag so we can drop the patch:

  https://todo.sr.ht/~mil/mepo-tickets/41

McSinyx commented:
Thanks!  GitHub UX kinda only allows merging all or nothing
(unless a maintainer pushes to your remote which is too spooky IMHO),
hence often quite some nitpick round trips are needed.

Commit:
mepo: add geoclueSupport?false

amjoseph-nixpkgs commented:
> Thanks! GitHub UX kinda only allows merging all or nothing

Ah, interesting; I did not know that.  I have plenty of complaints about github :)

> hence often quite some nitpick round trips are needed.

No problem.  Should be squashed now into four separate topic-specific commits at 8981daa3a8a8b325c0e331e0b972b0a109b3beb9.  This should make things easier down the line if one of them needs to be reverted.

McSinyx reviewed (approved):
Working as expected for me.

sikmir commented:
Result of `nixpkgs-review pr 189344` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>mepo</li>
    <li>mepo-x11</li>
  </ul>
</details>

sikmir commented:
I guess .orig file should not be installed:
```
$ ls results/mepo/bin/*.orig
results/mepo/bin/mepo_ui_menu_user_pin_updater.sh.orig
```

SuperSandro2000 commented on pkgs/applications/misc/mepo/default.nix:
```diff
@@ -16,26 +16,53 @@
 , withX11 ? false
+
+# transmits user's location without asking first
+, geoclueSupport ? false , geoclue2
+
+# transmits user's location without asking first
+, reportLocationToMozilla ? false
+
+, gpsdSupport ? true , gpsd
```
```suggestion
, geoclueSupport ? false, geoclue2

# transmits user's location without asking first
, reportLocationToMozilla ? false

, gpsdSupport ? true, gpsd
```
Shouldn't we assert something here? not sure what tbh

amjoseph-nixpkgs replied:
> Shouldn't we assert something here?

Why?

SuperSandro2000 replied:
Would it work to turn all on or off?

amjoseph-nixpkgs replied:
Correct: all 2^3 combinations of the three boolean flags are valid.

SuperSandro2000 commented on pkgs/applications/misc/mepo/default.nix:
```diff
@@ -16,26 +16,53 @@
+    ./dont-report-user-location-to-geoclue-without-asking.patch
+  ];
+
+  postPatch = lib.optionalString gpsdSupport ''
```
postPatch should be underneath src.

amjoseph-nixpkgs replied:
Implemented in 3e2f19b8a8a2

SuperSandro2000 commented on pkgs/applications/misc/mepo/default.nix:
```diff
@@ -16,26 +16,53 @@
+      scripts/mepo_ui_menu_user_pin_updater.sh \
+      --replace gpspipe ${gpsd}/bin/gpspipe
+  '' + lib.optionalString geoclueSupport ''
+    substituteInPlace \
+      scripts/mepo_ui_menu_user_pin_updater.sh \
+      --replace /usr/libexec/geoclue-2.0/demos/ \
+                ${geoclue2}/libexec/geoclue-2.0/demos/
```
```suggestion
    substituteInPlace scripts/mepo_ui_menu_user_pin_updater.sh \
      --replace /usr/libexec/geoclue-2.0/demos/ ${geoclue2}/libexec/geoclue-2.0/demos/
```

amjoseph-nixpkgs replied:
I think this is a case where the cost of sticking to 80 columns is low enough to be worthwhile.

SuperSandro2000 commented on pkgs/applications/misc/mepo/default.nix:
```diff
@@ -16,26 +16,53 @@
+    ++ lib.optional geoclueSupport geoclue2
+    ++ lib.optional gpsdSupport gpsd;
+
+  patches = lib.optionals (!reportLocationToMozilla) [
+    ./dont-report-user-location-to-mozilla-without-asking.patch
+  ] ++ lib.optionals (!geoclueSupport) [
+    ./dont-report-user-location-to-geoclue-without-asking.patch
+  ];
```
conditional patches are not great. Since this is just shell I would suggest to make this conditional based on an ENV or so and set that one for a wrapper.

amjoseph-nixpkgs replied:
Good point; I will revise.

amjoseph-nixpkgs replied:
On second thought: inventing our own environment variable for this is a bad idea.  When/if upstream gets around to responding to the bug filed about this being a bad idea for default behavior, they will create their own mechanism for opting in/out of this.  Then we'd either have to support the nixpkgs-specific environment variable forever (yuck) or have a breaking change with a really bad silent failure behavior.

So, I don't think inventing an environment variable is a great idea.

I think the only option for deconditionalizing the patches is to remove the options and apply the patches unconditionally.  This preserves the 0.4 behavior (never report the user's location to third parties).

I've implemented this in 926f3e6429a7.

McSinyx replied:
> conditional patches are not great

What possible issues may conditional patching cause?

amjoseph-nixpkgs replied:
> What possible issues may conditional patching cause?

If the patch is conditional and most users aren't building with that condition, nobody will notice if the patch ceases to apply cleanly.  For example, when the version is bumped.

So in general it is preferable to have any patches applied unconditionally -- when it's possible to do that.

Commit:
Update pkgs/applications/misc/mepo/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
https://github.com/NixOS/nixpkgs/pull/189344#discussion_r962339016

Commit:
remove geoclueSupport and reportLocationToMozilla flags

amjoseph-nixpkgs commented:
> I guess .orig file should not be installed:
> 
> ```
> $ ls results/mepo/bin/*.orig
> results/mepo/bin/mepo_ui_menu_user_pin_updater.sh.orig
> ```

I'm trying to reproduce this.

We might need `--no-backup-if-mismatch` in `patchFlags`.  I'm surprised that isn't the default.

amjoseph-nixpkgs commented:
So, based on upstream's response it doesn't look like there is an easy solution, and my concerns will likely recur in future version bumps.

I'll let the package maintainers @sikmir @McSinyx decide what to do here.  Feel free to use some/all/none of the changes in this PR.
